presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java

```diff
@@ -0,0 +1,471 @@
+                driverRunnerFactoriesWithTaskLifeCycle.add(new DriverSplitRunnerFactory(driverFactory, false));
+            }
+        }
+        this.driverRunnerFactoriesWithSplitLifeCycle = driverRunnerFactoriesWithSplitLifeCycle.build();
+        this.driverRunnerFactoriesWithTaskLifeCycle = driverRunnerFactoriesWithTaskLifeCycle.build();
+
+        checkArgument(this.driverRunnerFactoriesWithSplitLifeCycle.keySet().equals(tableScanSources),
+                "Fragment is partitioned, but not all partitioned drivers were found");
+
+        taskHandle = createTaskHandle(taskStateMachine, taskContext, localExecutionPlan, taskExecutor);
+    }
+
+    // this is a separate method to ensure that the `this` reference is not leaked during construction
+    private static TaskHandle createTaskHandle(
+            TaskStateMachine taskStateMachine,
+            TaskContext taskContext,
+            LocalExecutionPlan localExecutionPlan,
+            TaskExecutor taskExecutor)
+    {
+        TaskHandle taskHandle = taskExecutor.addTask(
+                taskStateMachine.getTaskId(),
+                () -> 0,
+                getInitialSplitsPerNode(taskContext.getSession()),
+                getSplitConcurrencyAdjustmentInterval(taskContext.getSession()),
+                getMaxDriversPerTask(taskContext.getSession()));
+        taskStateMachine.addStateChangeListener(state -> {
+            if (state.isDone()) {
+                taskExecutor.removeTask(taskHandle);
+                for (DriverFactory factory : localExecutionPlan.getDriverFactories()) {
+                    factory.noMoreDrivers();
+                }
+            }
+        });
+        return taskHandle;
+    }
+
+    public void start(List<TaskSource> sources)
+    {
+        requireNonNull(sources, "sources is null");
+
+        scheduleDriversForTaskLifeCycle();
+        scheduleDriversForSplitLifeCycle(sources);
+        checkTaskCompletion();
+    }
+
+    private void scheduleDriversForTaskLifeCycle()
+    {
+        List<DriverSplitRunner> runners = new ArrayList<>();
+        for (DriverSplitRunnerFactory driverRunnerFactory : driverRunnerFactoriesWithTaskLifeCycle) {
+            for (int i = 0; i < driverRunnerFactory.getDriverInstances().orElse(1); i++) {
+                runners.add(driverRunnerFactory.createDriverRunner(null));
+            }
+        }
+        enqueueDriverSplitRunner(true, runners);
+        for (DriverSplitRunnerFactory driverRunnerFactory : driverRunnerFactoriesWithTaskLifeCycle) {
+            driverRunnerFactory.noMoreDriverRunner();
+            verify(driverRunnerFactory.isNoMoreDriverRunner());
+        }
+    }
+
+    private synchronized void scheduleDriversForSplitLifeCycle(List<TaskSource> sources)
+    {
+        checkArgument(sources.stream().allMatch(TaskSource::isNoMoreSplits), "All task sources are expected to be final");
+
+        Map<PlanNodeId, List<ScheduledSplit>> splits = sources.stream()
```

Comment:
I am thinking if it makes sense to write imperative for-loop here? -- My general preference is when `Stream` API is used in such a complicated way, it's actually more difficult to understand than old good for-loop.

But it's just a personal taste.

Reply:
Yeah, i think it become way to cryptic.

Reply:
The iterative version + ListMultimap is much more readable